catalog/transformer/catalog-transformer-streaming-lib/src/main/java/org/codice/ddf/transformer/xml/streaming/lib/SaxEventHandlerDelegate.java

```diff
@@ -0,0 +1,378 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.xml.sax.Attributes;
+import org.xml.sax.ErrorHandler;
+import org.xml.sax.InputSource;
+import org.xml.sax.SAXException;
+import org.xml.sax.SAXParseException;
+import org.xml.sax.XMLReader;
+import org.xml.sax.helpers.DefaultHandler;
+import org.xml.sax.helpers.XMLReaderFactory;
+
+import ddf.catalog.data.Attribute;
+import ddf.catalog.data.Metacard;
+import ddf.catalog.data.MetacardType;
+import ddf.catalog.data.impl.AttributeImpl;
+import ddf.catalog.data.impl.BasicTypes;
+import ddf.catalog.data.impl.MetacardImpl;
+import ddf.catalog.transform.CatalogTransformerException;
+
+/**
+ * This class is used by the {@link XMLInputTransformer} to delegate SAX Parse events to the relevant handlers.
+ * Its primary method is {@link SaxEventHandlerDelegate#read} which takes in an {@link InputStream} returns a {@link Metacard},
+ * populated with all the {@link Attribute}s parsed by the {@link SaxEventHandlerDelegate#eventHandlers}
+ */
+public class SaxEventHandlerDelegate extends DefaultHandler {
+
+    private XMLReader parser;
+
+    private List<SaxEventHandler> eventHandlers = new ArrayList<>();
+
+    private Map<String, List<SaxEventHandler>> eventHandlerLookup = new HashMap<>();
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(SaxEventHandlerDelegate.class);
+
+    private MetacardType metacardType = BasicTypes.BASIC_METACARD;
+
+    public SaxEventHandlerDelegate() {
+        try {
+            parser = XMLReaderFactory.createXMLReader();
+        } catch (Exception e) {
+            LOGGER.debug(
+                    "Exception thrown during creation of SaxEventHandlerDelegate. Probably caused by one of the setFeature calls",
+                    e);
+        }
+    }
+
+    public SaxEventHandlerDelegate(SaxEventHandler... eventHandlers) {
+        this();
+        this.eventHandlers = Arrays.asList(eventHandlers);
+    }
+
+    public SaxEventHandlerDelegate(List<SaxEventHandler> eventHandlers) {
+        this();
+        this.eventHandlers = eventHandlers;
+    }
+
+    /**
+     * Takes in an {@link InputStream} returns a {@link Metacard},
+     * populated with all the {@link Attribute}s parsed by the {@link SaxEventHandlerDelegate#eventHandlers}
+     *
+     * @param inputStream an XML document that can be parsed into a Metacard
+     * @return a {@link Metacard},
+     * populated with all the {@link Attribute}s parsed by the {@link SaxEventHandlerDelegate#eventHandlers}
+     * @throws CatalogTransformerException
+     */
+    public Metacard read(InputStream inputStream) throws CatalogTransformerException {
+        configureEventHandlerLookup();
+
+        /*
+         * Create a new MetacardImpl with the proper MetacardType
+         */
+        Metacard metacard = new MetacardImpl(metacardType);
+
+        try {
+            InputSource newStream = new InputSource(new BufferedInputStream(inputStream));
```

Comment:
You don't always need the buffer, so sometimes you can gain an optimization benefit by skipping it.

The Java IO classes are an example of  the [decorator pattern](https://en.wikipedia.org/wiki/Decorator_pattern).